### PR TITLE
Create unit test for template

### LINKED ISSUE #9: Create unit test for template
## Why
Create unit test for the template to make sure that generated app can run properly.

## Acceptance Criteria
- Create unit test structure for the template.
- Add test cases for the template.
- All tests must pass on both local and CI.

### PR DESCRIPTION
**Pull request by gutakk**
## What happened 👀
Resolve #9 
- Add `Ginkgo` and `Gomega` to the template.
- Add Makefile and test command.
- Add template creation unit test cases.
- Update `create` command to receive `branch` flag (`main` is the default branch).
 
## Insight 📝
The idea of the template creation unit test is to run `cookiecutter` command to create an actual project and check every file that contains `{{cookiecutter.app_name}}` if they get replaced correctly.

Run test by
```sh
BRANCH=develop make test
```
 
## Proof Of Work 📹
All test must be passed
iYeHrEwdsF

### CODE REVIEW

File: tests/helpers.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		Fail("Failed to change directory: " + err.Error())
+	}
+}
+
+func CreateTemplate() string {
+	input := []byte("test-gin-templates")
+
+	reader, writer, err := os.Pipe()
+	if err != nil {
+		Fail("Failed to connected pair of files: " + err.Error())
```
Comment (olivierobert) [inline tests/helpers.go]: This error message is kind of weird 😆 When does that happen and what error it is?

File: tests/helpers.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	_, err = writer.Write(input)
+	if err != nil {
+		Fail("Failed to write file: " + err.Error())
+	}
+	writer.Close()
+
+	shCmd := exec.Command("cookiecutter", "../..")
```
Comment (olivierobert) [inline tests/helpers.go]: This should actually use the command `nimble-gin create` as would any user do. It's critical to use the built-in command.

File: tests/template/create_template_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = BeforeEach(func() {
+	tests.ChangeDirectory(TemplateGeneratedPath + "/test-gin-templates")
+})
+
+var _ = Describe("Create template", func() {
```
Comment (olivierobert) [inline tests/template/create_template_test.go]: To keep with the Go convention of keeping tests close to the source file, shouldn't this file be in `cmd` and named `create_test.go`?

File: tests/template/create_template_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	Context("given current directory", func() {
```
Comment (olivierobert) [inline tests/template/create_template_test.go]: How about:

```suggestion
	Context("given the project directory", func() {
```

File: tests/template/create_template_test.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		It("contains project name", func() {
+			dir, err := os.Getwd()
+			if err != nil {
+				Fail("Failed to get current directory: " + err.Error())
+			}
+
+			isContainProjectName := strings.Contains(dir, "gin-templates/tests/template/test-gin-templates")
```
Comment (olivierobert) [inline tests/template/create_template_test.go]: `isContainProjectName` is not syntactically correct. How about:

```suggestion
		directoryContainProjectName := strings.Contains(dir, "gin-templates/tests/template/test-gin-templates")
```

This variable would need to be different for each test though and it could be tedious e.g. `fileContainProjectName`

**Review by olivierobert — COMMENTED** [review]
The test are a bit light as they test only the presence of the project name. It can be enough for now but for future versions having an addon and variants structure like for the Rails and Elixir templates would allow breaking the tests into each addon.

File: tests/template/create_template_test.go (hunk 3 above)
Comment (gutakk) [reply tests/template/create_template_test.go]: Yes it should but I think it is better to separate in this case because in the future we will have addon and variants and they are part of `create` command. It is difficult to manage the test suite (the `BeforeSuite` and `AfterSuite` hooks) if we decide to breaking the tests into each addon.

What do you think?

File: tests/template/create_template_test.go (hunk 4 above)
Comment (gutakk) [reply tests/template/create_template_test.go]: Fixed in bf2e40575707b1ee2d9871cd1c51202e92e25fd9

File: tests/template/create_template_test.go (hunk 5 above)
Comment (gutakk) [reply tests/template/create_template_test.go]: Fixed in c9fd9e3f0cde380ae461289ef9e30c16b77fd178

File: tests/helpers.go (hunk 1 above)
Comment (gutakk) [reply tests/helpers.go]: Actually this command is for creating file to store stdin value.
Updated error message in ce7e2d0d9a1a7c4828c96a2db73ca20c70fad0f6

File: tests/helpers.go (hunk 2 above)
Comment (gutakk) [reply tests/helpers.go]: Yes it should but `nimble-gin create` doesn't allow us to test on current working branch but only default git branch because `create` command executes `cookiecutter https://github.com/nimblehq/gin-templates` which use source code from default git branch.

I cannot think of the best way to done this. Do you have any idea?

File: tests/template/create_template_test.go (hunk 3 above)
Comment (olivierobert) [reply tests/template/create_template_test.go]: As we are not sure how the file will be structured in that case (addons might be sub-commands for instans), it seems too early to make the decision to separate it now. So I recommend keeping this file in `cmd` as it fits the current architecture.

File: tests/helpers.go (hunk 2 above)
Comment (olivierobert) [reply tests/helpers.go]: The branch name should be a parameter in the command and default to the master branch if it's omitted: 

image

cookie-cutter supports specifying a branch (see [docs](https://cookiecutter.readthedocs.io/en/1.7.2/usage.html#works-directly-with-git-and-hg-mercurial-repos-too)):

image

File: tests/template/create_template_test.go (hunk 3 above)
Comment (gutakk) [reply tests/template/create_template_test.go:17]: Agree! Fixed in eb0eed87b0e69e7443d7c753a9898423a487a552

Comment (gutakk) [reply tests/template/create_template_test.go]: Fixed in ce89d2babcc1db8397def10361acf21b66e633c4

File: tests/helpers.go (hunk 2 above)
Comment (gutakk) [reply tests/helpers.go]: Fixed in d5b18bec99962908241262f937ed088039a1e860 c5e2d4ddce83a3b60fc5edad31db906366c77eb9

Comment (gutakk) [reply tests/helpers.go]: To test if the application template has new tests.

- Checkout to `test-template-unit-test` branch.
- Run `BRANCH=develop make test`. This should fail because template test cannot find `test-specifying-branch.yml`
- Run `BRANCH=test-template-unit-test make test`. This should pass.

File: README.md [hunk 6]
```diff
@@ -26,15 +26,26 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
...
-  go build -i -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
+  go build -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
   ```
-- Run create the project command.
+- Create the project from a template on a specifying branch.
   ```
-  nimble-gin create
+  nimble-gin -b {{git branch}} create
   ```
+  **Note:** `-b` can be omitted, branch will be `main` by default.
```
Comment (olivierobert) [inline README.md]: I recommend keeping the general use case by default in the README. Then in the wiki, there can be additional options detailed. So I recommend to revert this change and add wiki page to explain all options.

File: README.md [hunk 7]
```diff
@@ -26,15 +26,26 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
   ```
 - Build the binary file.
   ```
-  go build -i -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
+  go build -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
   ```
-- Run create the project command.
+- Create the project from a template on a specifying branch.
```
Comment (olivierobert) [inline README.md]: Not correct grammatically:

```suggestion
- Create the project from a template on a specific branch.
```

File: README.md [hunk 8]
```diff
@@ -26,15 +26,26 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
...
+  **Note:** `-b` can be omitted, branch will be `main` by default.
 - Follow the instructions in the terminal.
 - Your new application is created 🎉 .
 
+## Test
+
+Execute all unit tests:
+
+```sh
+BRANCH={{git branch}} make test
```
Comment (olivierobert) [inline README.md]: Just like above, the README should focus on the common use case. This can be moved to the wiki.

File: tests/template.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+	currentPath, err := os.Getwd()
+	if err != nil {
+		Fail("Failed to get current directory: " + err.Error())
+	}
+
+	return currentPath
+}
+
+func DownloadGinTemplate() {
```
Comment (olivierobert) [inline tests/template.go]: Just minor preference to order the methods of this file ordered by the use case:

```go
func DownloadGinTemplate() {}
func BuildGinTemplate() {}
func RemoveCookiecuttersCache() {}
func CreateProjectFromGinTemplate() string {}
```

As a general rule, prefer ordering methods how they are used logically, not how/when the methods were added. It's better for code discoverability, readability and maintenance.

File: cmd/create.go [hunk 10]
```diff
@@ -24,23 +24,28 @@ import (
 )
 
 // createCmd represents the create command
```
Comment (olivierobert) [inline cmd/create.go]: How about rephrasing this to:

```suggestion
// createCmd generates the project using cookiecutter
```

**Review by olivierobert — COMMENTED** [review]
Looks better now 💪 A few minor changes needed and then good to go.

File: tests/template.go (hunk 9 above)
Comment (gutakk) [reply tests/template.go]: Fixed in 146811705e93b2b23a0b069ce136ee88f740c1e2

File: cmd/create.go (hunk 10 above)
Comment (gutakk) [reply cmd/create.go]: Fixed in e2cf98e3c0ddd8281cca1f29aca709577fdbaee4. 
Actually, that comment was generated by Cobra.

File: README.md (hunk 6 above)
Comment (gutakk) [reply README.md]: Fixed in 2a50e67df6f4112a5236c6eba238a736dacc2df9.
I created a new issue to add command's flags information in https://github.com/nimblehq/gin-templates/issues/26

File: README.md (hunk 7 above)
Comment (gutakk) [reply README.md]: Fixed in 2a50e67df6f4112a5236c6eba238a736dacc2df9.
I think if we decide to keeping general use case then specific/specifying should be changed to `main` (which is the default git branch)

File: README.md (hunk 8 above)
Comment (gutakk) [reply README.md]: Fixed in 2a50e67df6f4112a5236c6eba238a736dacc2df9.
I created a new issue to add command's flags information in https://github.com/nimblehq/gin-templates/issues/26

File: cmd/create_test.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+var _ = Describe("Create template", func() {
+	Context("given the project directory", func() {
+		It("contains project name", func() {
+			dir, err := os.Getwd()
+			if err != nil {
+				Fail("Failed to get current directory: " + err.Error())
+			}
+
+			directoryContinProjectName := strings.Contains(dir, "gin-templates/cmd/test-gin-templates")
```
Comment (junan) [inline cmd/create_test.go]:
```suggestion
			directoryContainProjectName := strings.Contains(dir, "gin-templates/cmd/test-gin-templates")
```

File: cmd/create.go [hunk 12]
```diff
@@ -23,24 +23,29 @@ import (
...
+		Run: func(cmd *cobra.Command, args []string) {
+			shCmd := exec.Command("cookiecutter", "https://github.com/nimblehq/gin-templates.git", "--checkout", branchFlag)
+			shCmd.Stdout = os.Stdout
+			shCmd.Stdin = os.Stdin
+
+			err := shCmd.Run()
+			if err != nil {
+				log.Fatal("Gin project created unsuccessfully: ", err)
+			} else {
+				log.Print("Gin project created successfully.")
```
Comment (junan) [inline cmd/create.go:42]: We used `Println` olthers places, should we use `Println` use it here also?

File: README.md [hunk 13]
```diff
@@ -26,15 +26,23 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
   ```
 - Build the binary file.
   ```
-  go build -i -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
+  go build -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
   ```
-- Run create the project command.
+- Create the project from a template on a **main** branch.
   ```
```
Comment (junan) [inline README.md:36]: Can we add an empty line before it?

File: cmd/create.go (hunk 12 above)
Comment (junan) [reply cmd/create.go:42]: Not sure if it is the last output or so, if it is then ignore this comment

File: cmd/create_test.go (hunk 11 above)
Comment (gutakk) [reply cmd/create_test.go]: Good catch. Fixed in 8a02884b1579d1ef7d9c70e69a90e5099de6a780

File: README.md (hunk 13 above)
Comment (gutakk) [reply README.md:36]: Sure. Fixed in e6e50849fc249d06e49250271c2d33233b3f1001

**Review by junan — APPROVED** [review]

File: README.md [hunk 14]
```diff
@@ -12,29 +12,43 @@ Our templates offer a rich boilerplate to jump start Go Gin-based application de
...
+
 - Build the binary file.
   ```
-  go build -i -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
+  go build -o $GOPATH/bin/nimble-gin github.com/nimblehq/gin-templates
   ```
-- Run create the project command.
+
+- Create the project from a template on a **main** branch.
```
Comment (olivierobert) [inline README.md]: English-wise, it still could be better:

```suggestion
- Create the project using gin-templates. Note that the **main** branch is used.
```

Later on, when the wiki is added, the content could be enhanced with thee following:

```suggestion
- Create the project using gin-templates. Note that the **main** branch is used by default. Refer to [this wiki page](REPLACE WITH LINK TO WIKI PAGE) for instructions on how to use a different branch.
```

**Review by olivierobert — CHANGES_REQUESTED** [review]

File: README.md (hunk 14 above)
Comment (gutakk) [reply README.md]: Fixed in f025cbc28da3533a347467ed4e806b511ee97c7c

**Review by olivierobert — APPROVED** [review]
